### DISCUSSION TITLE
Receiving the final sender

### DISCUSSION THREAD

**Discussion by Ender-Fake** [PaperMC/Paper#11344] [Ideas]
### Is your feature request related to a problem?

**CommandSourceStack** has **CommandSender** and **Executor** but when you want to get the last sender you have to get it yourself

### Describe the solution you'd like.

I would like to have a method that returns the last sender

### Describe alternatives you've considered.

Nothing

### Other

At the moment I'm using this
```
public void execute(@NotNull CommandSourceStack commandSourceStack, @NotNull String[] args) {
    CommandSender sender = commandSourceStack.getSender();
    if (commandSourceStack.getExecutor()!=null) sender = commandSourceStack.getExecutor();
    /* code */
}
```

**Comment by Machine-Maker** [comment]
Yes, I agree that such a method should be used, and it should be pointed to as the "preferred" method to use. In fact, it might be best to change one of the existing methods to that since the API is still experimental.

**Reply by Ender-Fake** [reply]
If you replace the method, you can add another method to check if the executor has been changed.

**Comment by willkroboth** [comment]
Maybe the `getExecutor` method could be renamed to `getProxyEntity`/`getTargetEntity`, with javadocs that say something like, `The entity targeted by this command, which can be changed using "/execute as <entity> run <command>"`. If `execute as` was not used to target an entity, `getProxyEntity` would default to the Player who sent the command, or `null` if the console or a command block ran the command. There could also be a `boolean isProxySender()` method like Ender-Fake mentioned to check if the target entity was changed using `execute as`.

That might help make it clear that `getSender` should be used for checking permissions, while `getProxyEntity` should be used for performing actions if that is what the plugin wants to do.
